library/spdm_common_lib/libspdm_com_opaque_data.c

```diff
@@ -192,6 +191,133 @@ libspdm_build_opaque_data_supported_version_data(libspdm_context_t *spdm_context
+    if ((data_in_size == 0) || (data_in == NULL)) {
+        return false;
+    }
+
+    if (libspdm_get_connection_version (spdm_context) >= SPDM_MESSAGE_VERSION_12) {
+        spdm_general_opaque_data_table_header = data_in;
+        if (spdm_general_opaque_data_table_header->total_elements < 1) {
+            return LIBSPDM_STATUS_INVALID_MSG_FIELD;
+        }
+        opaque_element_table_header =
+            (const void *)(spdm_general_opaque_data_table_header + 1);
+
+        element_num = spdm_general_opaque_data_table_header->total_elements;
+        if (data_in_size < sizeof(spdm_general_opaque_data_table_header_t)) {
+            return false;
+        }
+        data_element_size = data_in_size - sizeof(spdm_general_opaque_data_table_header_t);
+    } else {
+        general_opaque_data_table_header = data_in;
+        if ((general_opaque_data_table_header->spec_id !=
+             SECURED_MESSAGE_OPAQUE_DATA_SPEC_ID) ||
+            (general_opaque_data_table_header->opaque_version !=
+             SECURED_MESSAGE_OPAQUE_VERSION) ||
+            (general_opaque_data_table_header->total_elements < 1)) {
+            return LIBSPDM_STATUS_INVALID_MSG_FIELD;
+        }
+        opaque_element_table_header =
+            (const void *)(general_opaque_data_table_header + 1);
+
+        element_num = general_opaque_data_table_header->total_elements;
+        if (data_in_size < sizeof(secured_message_general_opaque_data_table_header_t)) {
+            return false;
+        }
+        data_element_size = data_in_size -
+                            sizeof(secured_message_general_opaque_data_table_header_t);
+    }
+
+    for (element_index = 0; element_index < element_num;
+         element_index++) {
+
+        /*check element header id*/
+        if (opaque_element_table_header->id > SPDM_REGISTRY_ID_JEDEC) {
+            return false;
+        }
+
+        current_element_len = sizeof(secured_message_opaque_element_table_header_t) +
+                              opaque_element_table_header->opaque_element_data_len;
+        /* Add Padding*/
+        current_element_len = (current_element_len + 3) & ~3;
+
+        total_element_len += current_element_len;
+
+        if (data_element_size < total_element_len) {
+            return false;
+        }
+
+        if (opaque_element_table_header->id == element_id) {
+            secured_message_element_header =
+                (const void *)(opaque_element_table_header + 1);
+            if ((secured_message_element_header->sm_data_id == sm_data_id) ||
+                (secured_message_element_header->sm_data_version ==
+                 SECURED_MESSAGE_OPAQUE_ELEMENT_SMDATA_DATA_VERSION)) {
+                /*get element by element id*/
+                *get_element_ptr = opaque_element_table_header;
+                *get_element_len = current_element_len;
+                result = true;
+            }
+        }
+
+        /*move to next element*/
+        opaque_element_table_header =
```

Comment:
How do you know this new opaque_element_table_header is valid?

Reply:
The problem is that you move to the next, but you did not check the size.

The pointer may already out of range, before you deference it.